metrics.go

```diff
@@ -0,0 +1,431 @@
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+
+	"go.opencensus.io/metric/metricdata"
+	"go.opencensus.io/resource"
+)
+
+var (
+	errLableExtraction       = errors.New("error extracting labels")
+	errUnspecifiedMetricKind = errors.New("metric kind is unpsecified")
+)
+
+// ExportMetrics exports OpenCensus Metrics to Stackdriver Monitoring.
+func (se *statsExporter) ExportMetrics(ctx context.Context, metrics []*metricdata.Metric) error {
+	if len(metrics) == 0 {
+		return errNilMetric
+	}
+
+	for _, metric := range metrics {
+		se.protoMetricsBundler.Add(metric, 1)
+		// TODO: [rghetia] handle errors.
+	}
+
+	return nil
+}
+
+func (se *statsExporter) handleMetricsUpload(metrics []*metricdata.Metric) {
+	err := se.uploadMetrics(metrics)
+	if err != nil {
+		se.o.handleError(err)
+	}
+}
+
+func (se *statsExporter) uploadMetrics(metrics []*metricdata.Metric) error {
+	ctx, cancel := se.o.newContextWithTimeout()
+	defer cancel()
+
+	ctx, span := trace.StartSpan(
+		ctx,
+		"contrib.go.opencensus.io/exporter/stackdriver.uploadMetrics",
+		trace.WithSampler(trace.NeverSample()),
+	)
+	defer span.End()
+
+	for _, metric := range metrics {
+		// Now create the metric descriptor remotely.
+		if err := se.createMetricDescriptorFromMetric(ctx, metric); err != nil {
+			span.SetStatus(trace.Status{Code: 2, Message: err.Error()})
+			return err
+		}
+	}
+
+	var allTimeSeries []*monitoringpb.TimeSeries
+	for _, metric := range metrics {
+		tsl, err := se.metricToMpbTs(ctx, metric)
+		if err != nil {
+			span.SetStatus(trace.Status{Code: 2, Message: err.Error()})
+			return err
+		}
+		allTimeSeries = append(allTimeSeries, tsl...)
+	}
+
+	// Now batch timeseries up and then export.
+	for start, end := 0, 0; start < len(allTimeSeries); start = end {
+		end = start + maxTimeSeriesPerUpload
+		if end > len(allTimeSeries) {
+			end = len(allTimeSeries)
+		}
+		batch := allTimeSeries[start:end]
+		ctsreql := se.combineTimeSeriesToCreateTimeSeriesRequest(batch)
+		for _, ctsreq := range ctsreql {
+			if err := createTimeSeries(ctx, se.c, ctsreq); err != nil {
+				span.SetStatus(trace.Status{Code: trace.StatusCodeUnknown, Message: err.Error()})
+				// TODO(@rghetia): Don't fail fast here, perhaps batch errors?
+				// return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// metricToMpbTs converts a metric into a Stackdriver Monitoring v3 API CreateTimeSeriesRequest
+// but it doesn't invoke any remote API.
+func (se *statsExporter) metricToMpbTs(ctx context.Context, metric *metricdata.Metric) ([]*monitoringpb.TimeSeries, error) {
+	if metric == nil {
+		return nil, errNilMetric
+	}
+
+	resource := metricRscToMpbRsc(metric.Resource)
+
+	metricName := metric.Descriptor.Name
+	metricType, _ := se.metricTypeFromProto(metricName)
+	metricLabelKeys := metric.Descriptor.LabelKeys
+	metricKind, _ := metricDescriptorTypeToMetricKind(metric)
+
+	if metricKind == googlemetricpb.MetricDescriptor_METRIC_KIND_UNSPECIFIED {
+		return nil, errUnspecifiedMetricKind
```

Comment:
Same here, I don't think this is an error, otherwise we won't be able to handle the `Summary` type.